x-pack/plugins/security_solution/public/detection_engine/rule_creation/logic/esql_query_columns.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { QueryClient, QueryFunction, QueryKey } from '@tanstack/react-query';
+import { useQuery } from '@tanstack/react-query';
+import type { DatatableColumn } from '@kbn/expressions-plugin/common';
+import { getESQLQueryColumns } from '@kbn/esql-utils';
+import { KibanaServices } from '../../../common/lib/kibana';
+
+const DEFAULT_STALE_TIME = 60 * 1000;
+
+interface FetchEsqlQueryColumnsParams {
+  esqlQuery: string;
+  queryClient: QueryClient;
+}
+
+export async function fetchEsqlQueryColumns({
+  esqlQuery,
+  queryClient,
+}: FetchEsqlQueryColumnsParams): Promise<DatatableColumn[]> {
+  const data = await queryClient.fetchQuery({
+    queryKey: [esqlQuery.trim()],
+    queryFn: queryEsqlColumnsFactory(esqlQuery),
+    staleTime: DEFAULT_STALE_TIME,
+    retry: false,
+  });
+
+  if (data instanceof Error) {
+    throw data;
+  }
+
+  return data;
+}
+
+interface UseEsqlQueryColumnsResult {
+  columns: DatatableColumn[];
+  isLoading: boolean;
+}
+
+export function useEsqlQueryColumns(esqlQuery: string): UseEsqlQueryColumnsResult {
```

Comment:
`fetchEsqlQueryColumns` and `useEsqlQueryColumns` have different use cases. The only reason we don't fetch columns directly is caching React Query provides. These two functions follow minimum interface principle where only necessary data is exposed. `fetchEsqlQueryColumns`  use case requires distinguishing between error and `_id` column missing cases. Throwing exceptions helps here.

Usual practice is that `fetch` throws while hooks usually don't throw errors but rather return an (expected) error so it could be handled. Error boundary is usually used to catch unexpected errors.

Right now we don't handle errors in UI in any special way like showing a toast message. If we decided to address in in the future it's possible to handle an error inside a hook and expose an error to be consumed outside.